Sources/QuickCheck/ArbitraryDecodable.swift

```diff
@@ -0,0 +1,218 @@
+      self.codingPath = decoder.codingPath
+      self.decoder = decoder
+    }
+
+    func contains(_ key: Key) -> Bool {
+      return true
+    }
+
+    func decodeNil(forKey key: Key) throws -> Bool {
+      return self.decoder.run(uniform.map { $0 >= 0.75 })
+    }
+
+    func decode<T>(_ type: T.Type, forKey key: Key) throws -> T where T: Decodable {
+      self.decoder.codingPath.append(key)
+      defer { self.decoder.codingPath.removeLast() }
+      return try T(from: self.decoder)
+    }
+
+    func nestedContainer<NestedKey>(keyedBy type: NestedKey.Type, forKey key: Key) throws
+      -> KeyedDecodingContainer<NestedKey> where NestedKey: CodingKey {
+        self.decoder.codingPath.append(key)
+        defer { self.decoder.codingPath.removeLast() }
+        return .init(KeyedContainer<NestedKey>(decoder: self.decoder))
+    }
+
+    func nestedUnkeyedContainer(forKey key: Key) throws -> UnkeyedDecodingContainer {
+      self.decoder.codingPath.append(key)
+      defer { self.decoder.codingPath.removeLast() }
+      return UnkeyedContainer(decoder: self.decoder)
+    }
+
+    func superDecoder() throws -> Decoder {
+      return self.decoder
+    }
+
+    func superDecoder(forKey key: Key) throws -> Decoder {
+      self.decoder.codingPath.append(key)
+      defer { self.decoder.codingPath.removeLast() }
+      return self.decoder
+    }
+  }
+
+  struct UnkeyedContainer: UnkeyedDecodingContainer {
+    let codingPath: [CodingKey]
+    let count: Int?
+    var currentIndex = 0
+    let decoder: ArbitraryDecoder
+    var isAtEnd: Bool { return currentIndex == count }
+
+    init(decoder: ArbitraryDecoder) {
+      self.codingPath = decoder.codingPath
+      print(decoder.genState.size)
+      self.count = decoder.run(choose(0..<decoder.genState.size) |> suchThat <| { $0 % 2 == 0 })
+      self.decoder = decoder
+    }
+
+    func decodeNil() throws -> Bool {
+      return self.decoder.run(uniform.map { $0 >= 0.75 })
+    }
+
+    func decode<T>(_ type: T.Type) throws -> T where T: Decodable {
+      return try T(from: self.decoder)
+    }
+
+    func nestedContainer<NestedKey>(keyedBy type: NestedKey.Type) throws
+      -> KeyedDecodingContainer<NestedKey> where NestedKey: CodingKey {
+        return .init(KeyedContainer<NestedKey>(decoder: self.decoder))
+    }
+
+    func nestedUnkeyedContainer() throws -> UnkeyedDecodingContainer {
+      return self
+    }
+
+    mutating func superDecoder() throws -> Decoder {
+      defer { self.currentIndex += 1 }
+      return self.decoder
+    }
+  }
+
+  struct SingleValueContainer: SingleValueDecodingContainer {
+    let codingPath: [CodingKey]
+    let decoder: ArbitraryDecoder
+
+    init(decoder: ArbitraryDecoder) {
+      self.codingPath = decoder.codingPath
+      self.decoder = decoder
+    }
+
+    func decodeNil() -> Bool {
+      return self.decoder.run(uniform.map { $0 >= 0.75 })
+    }
+
+    func decode(_ type: Bool.Type) throws -> Bool {
+      return self.decoder.run(genBool)
```

Comment:
The latest Swift snapshot lets us define these base type decodings in one place! When I tried on beta 3 I had to reproduce these in each other decoding container, twice! :(